src/metric.cc

```diff
@@ -0,0 +1,324 @@
+const MetricRequestKind&
+Metric::RequestKind()
+{
+  return metric_request_kind_;
+}
+
+double
+Metric::Value()
+{
+  return value_;
+}
+
+void
+Metric::SaveToSharedMemory(std::unique_ptr<SharedMemoryManager>& shm_pool)
+{
+  AllocatedSharedMemory<char> custom_metric_shm =
+      shm_pool->Construct<char>(sizeof(MetricShm), true);
+  custom_metric_shm_ptr_ =
+      reinterpret_cast<MetricShm*>(custom_metric_shm.data_.get());
+
+  std::unique_ptr<PbString> family_name_shm =
+      PbString::Create(shm_pool, FamilyName());
+  std::unique_ptr<PbString> labels_shm = PbString::Create(shm_pool, Labels());
+
+  custom_metric_shm_ptr_->value = value_;
+  custom_metric_shm_ptr_->metric_request_kind = metric_request_kind_;
+  custom_metric_shm_ptr_->family_name_shm_handle = family_name_shm->ShmHandle();
+  custom_metric_shm_ptr_->labels_shm_handle = labels_shm->ShmHandle();
+  custom_metric_shm_ptr_->metric_family_address = metric_family_address_;
+  custom_metric_shm_ptr_->metric_address = metric_address_;
+
+  // Save the references to shared memory.
+  custom_metric_shm_ = std::move(custom_metric_shm);
+  family_name_shm_ = std::move(family_name_shm);
+  labels_shm_ = std::move(labels_shm);
+  shm_handle_ = custom_metric_shm.handle_;
+}
+
+std::unique_ptr<Metric>
+Metric::LoadFromSharedMemory(
+    std::unique_ptr<SharedMemoryManager>& shm_pool,
+    bi::managed_external_buffer::handle_t handle)
+{
+  AllocatedSharedMemory<char> custom_metric_shm = shm_pool->Load<char>(handle);
+  MetricShm* custom_metric_shm_ptr =
+      reinterpret_cast<MetricShm*>(custom_metric_shm.data_.get());
+
+  std::unique_ptr<PbString> family_name_shm = PbString::LoadFromSharedMemory(
+      shm_pool, custom_metric_shm_ptr->family_name_shm_handle);
+  std::unique_ptr<PbString> labels_shm = PbString::LoadFromSharedMemory(
+      shm_pool, custom_metric_shm_ptr->labels_shm_handle);
+
+  return std::unique_ptr<Metric>(
+      new Metric(custom_metric_shm, family_name_shm, labels_shm));
+}
+
+Metric::Metric(
+    AllocatedSharedMemory<char>& custom_metric_shm,
+    std::unique_ptr<PbString>& family_name_shm,
+    std::unique_ptr<PbString>& labels_shm)
+    : custom_metric_shm_(std::move(custom_metric_shm)),
+      family_name_shm_(std::move(family_name_shm)),
+      labels_shm_(std::move(labels_shm))
+{
+  custom_metric_shm_ptr_ =
+      reinterpret_cast<MetricShm*>(custom_metric_shm_.data_.get());
+  family_name_ = family_name_shm_->String();
+  labels_ = labels_shm_->String();
+  value_ = custom_metric_shm_ptr_->value;
+  metric_request_kind_ = custom_metric_shm_ptr_->metric_request_kind;
+  metric_family_address_ = custom_metric_shm_ptr_->metric_family_address;
+  metric_address_ = custom_metric_shm_ptr_->metric_address;
+}
+
+void*
+Metric::MetricAddress()
+{
+  return metric_address_;
+}
+
+#ifdef TRITON_PB_STUB
+void
+Metric::SendIncrementRequest(const double& value)
+{
+  std::unique_ptr<Stub>& stub = Stub::GetOrCreateInstance();
+  metric_request_kind_ = MetricIncrement;
```

Comment:
I think `metric_request_kind_` should not be a part of this object. It should be the command field of the `IPCMessage`.